### PR TITLE
Ensure HttpListener request buffer is aligned as required by the host processor

### LINKED ISSUE #25289: HttpListener failure on Windows ARM
System.Net.HttpListenerException: Invalid access to memory location
   at System.Net.HttpListener.BeginGetContext(AsyncCallback callback, Object state)
   at DotNetCoreHttpListener.Program.Main(String[] args) in C:\src\DotNetCoreHttpListener\Program.cs:line 22

### PR DESCRIPTION
**Pull request by rmkerr**
On Windows, the HttpRecieveHttpRequest function requires a buffer with a memory alignment greater than or equal to the required alignment of the HTTP_REQUEST struct. 

This fix ensures that alignment requirements are respected when allocating buffers in HttpListener RequestContextBase. Since HttpReceiveHttpRequest copies both the HTTP_REQUEST struct and the variable length request body into the buffer, we need to be able to allocate a buffer with variable size and with a set alignment. Since C# does not provide a method for specifying the alignment of byte arrays, I switched the underlying buffer to be unmanaged. This unmanaged buffer is allocated using Marshal.AllocHGlobal, which allocates memory at the maximum alignment required by the host processor.

I'm new to writing code on the border between managed and unmanaged, and so I would appreciate any suggestions on how to improve things here. 

Fixes #25289

### CODE REVIEW

File: src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs [hunk 1]
```diff
@@ -786,75 +786,70 @@ internal static unsafe string GetVerb(HTTP_REQUEST* request)
             return GetVerb(request, 0);
         }
 
-        internal static unsafe string GetVerb(byte[] memoryBlob, IntPtr originalAddress)
+        internal static unsafe string GetVerb(IntPtr memoryBlob, IntPtr originalAddress)
         {
-            fixed (byte* pMemoryBlob = memoryBlob)
```
Comment (rmkerr) [inline src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs]: The vast majority of the line changes in this file are the result of removing these fixed statements, and thus changing the indentation for large blocks of code.

File: src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs [hunk 2]
```diff
@@ -786,75 +786,70 @@ internal static unsafe string GetVerb(HTTP_REQUEST* request)
...
-                        headerCollection.Add(HTTP_REQUEST_HEADER_ID.ToString(index), headerValue);
-                    }
-                    pKnownHeader++;
+                    string headerValue = new string(pKnownHeader->pRawValue + fixup, 0, pKnownHeader->RawValueLength);
+                    headerCollection.Add(HTTP_REQUEST_HEADER_ID.ToString(index), headerValue);
                 }
+                pKnownHeader++;
             }
-
+
```
Comment (davidsh) [inline src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs]: nit: remove extra tab/spacing.

File: src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs (hunk 1 above)
Comment (stephentoub) [reply src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs]: > The vast majority of the line changes in this file are the result of removing these fixed statements, and thus changing the indentation for large blocks of code

FYI, if you append ?w=1 to the GitHub url, it'll ignore whitespace differences:
https://github.com/dotnet/corefx/pull/25563/files?w=1

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs [hunk 3]
```diff
@@ -116,7 +125,19 @@ protected void UnsetBlob()
...
-            _backingBuffer = size == 0 ? null : new byte[size];
+            if(_backingBuffer != IntPtr.Zero)
+            {
+                Marshal.FreeHGlobal(_backingBuffer);
+            }
+
+            _backingBuffer = size == 0 ? IntPtr.Zero : Marshal.AllocHGlobal(size);
+            _backingBufferLength = size;
+
+            // Zero out the contents of the buffer.
```
Comment (rmkerr) [inline src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: The buffer needs to be zero'd, which is done by default for managed memory. I'd like to do this faster, but I don't know what the best way is.

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs [hunk 4]
```diff
@@ -48,7 +50,14 @@ public void Dispose()
             Dispose(true);
         }
 
-        protected virtual void Dispose(bool disposing) { }
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_backingBuffer != IntPtr.Zero)
```
Comment (davidsh) [inline src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: Please review IDisposable pattern.  Not sure if you should add an 'if' check for "disposing == true" here.

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs [hunk 5]
```diff
@@ -116,7 +125,19 @@ protected void UnsetBlob()
...
+            if(_backingBuffer != IntPtr.Zero)
+            {
+                Marshal.FreeHGlobal(_backingBuffer);
+            }
+
+            _backingBuffer = size == 0 ? IntPtr.Zero : Marshal.AllocHGlobal(size);
+            _backingBufferLength = size;
+
+            // Zero out the contents of the buffer.
+            for(int i = 0; i < size; ++i)
```
Comment (davidsh) [inline src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: nit: add space character after 'for'

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs [hunk 6]
```diff
@@ -116,7 +125,19 @@ protected void UnsetBlob()
 
         protected void SetBuffer(int size)
         {
-            _backingBuffer = size == 0 ? null : new byte[size];
+            if(_backingBuffer != IntPtr.Zero)
```
Comment (davidsh) [inline src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: nit: add space character after 'if'

File: src/System.Net.HttpListener/src/System/Net/Windows/SyncRequestContext.cs [hunk 7]
```diff
@@ -9,56 +9,27 @@ namespace System.Net
...
-                    return RequestBlob;
-                }
-                _pinnedHandle.Free();
-            }
-            SetBuffer(size);
-            if (RequestBuffer == null)
+            if (Size > 0 && Size == newSize)
             {
-                return null;
+                 return RequestBlob;
```
Comment (davidsh) [inline src/System.Net.HttpListener/src/System/Net/Windows/SyncRequestContext.cs]: spacing looks wrong

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs (hunk 4 above)
Comment (rmkerr) [reply src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: I think that the current behavior is correct, based on [this page](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#dispose-and-disposeboolean).

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs [hunk 8]
```diff
@@ -116,7 +125,19 @@ protected void UnsetBlob()
...
+            }
+
+            _backingBuffer = size == 0 ? IntPtr.Zero : Marshal.AllocHGlobal(size);
+            _backingBufferLength = size;
+
+            // Zero out the contents of the buffer.
+            for(int i = 0; i < size; ++i)
+            {
+                Marshal.WriteByte(_backingBuffer + i, 0);
+            }
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: How big are these usually?  Calling Marshal.WriteByte per byte for a large buffer isn't cheap.  Alternatively you could use span:
```C#
new Span<byte>((byte*)_backingBuffer, 0, size).Fill(0);
```

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs (hunk 4 above)
Comment (davidsh) [reply src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: ok.  What about a 'finalizer'/  Does this have one already and do we need one.  Reading the pattern description from your link above says this:

>You implement the IDisposable interface and an additional Dispose(Boolean) method, and you also override the Object.Finalize method. You must override Finalize to ensure that unmanaged resources are disposed of if your IDisposable.Dispose implementation is not called by a consumer of your type. If you use the recommended technique discussed in the previous bullet, the System.Runtime.InteropServices.SafeHandle class does this on your behalf.

File: src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs (hunk 1 above)
Comment (rmkerr) [reply src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs]: Good to know!

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs (hunk 8 above)
Comment (rmkerr) [reply src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: The default size is 4096, but they can be larger than that depending on the size of the request. I really don't like having that loop there, so I'll look into doing it that way instead. Thanks!

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs (hunk 4 above)
Comment (rmkerr) [reply src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: RequestContextBase does implement a finalizer that calls Dispose(bool), freeing the unmanaged resources it has acquired. I'm not very familiar with the pattern though, so I'll read through the rest of that dispose/finalize article and make sure I'm doing everything correctly.

**Comment by davidsh** [comment]
@rmkerr Please run outerloop tests as well since this is required for any networking code changes.

File: src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs (hunk 2 above)
Comment (davidsh) [reply src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs]: I'm still seeing this extra space even in your latest set of commits.

**Comment by rmkerr** [comment]
@davidsh HttpListener outerloop tests are passing.

**Comment by davidsh** [comment]
>@davidsh HttpListener outerloop tests are passing.

You need to run the outerloop tests in the PR itself and not just your dev machine.

**Comment by rmkerr** [comment]
Like this?
@dotnet-bot test Outerloop Windows x64 Debug Build please

**Comment by davidsh** [comment]
>Like this?
@dotnet-bot test Outerloop Windows x64 Debug Build please

Yes, and you should run all platforms affected. And it's probably a good idea to run Linux as well just to be sure that your changes don't impact that platform.

**Comment by rmkerr** [comment]
@dotnet-bot test Outerloop Windows x86 Debug Build please

The change should be windows only, but just in case:
@dotnet-bot test Outerloop Linux x64 Debug Build please
@dotnet-bot test Outerloop OSX x64 Debug Build please

**Comment by rmkerr** [comment]
I got failures in the following outerloop tests for HttpClient:

> HttpClient_ClientUsesAuxRecord_Ok
> ReadAsStreamAsync_ValidServerResponse_Success(transferType: ContentLength, transferError: None)
> GetAsync_MaxLimited_ConcurrentCallsStillSucceed(maxConnections: 3, numRequests: 2, secure: True)

I took a look at the tests, and I don't believe that these failures are related. Especially since the code is only changing HttpListener, which is not used by HttpClient. Just to get more info I'm running Outerloop again.

@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Windows x86 Debug Build please

File: src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs (hunk 4 above)
Comment (rmkerr) [reply src/System.Net.HttpListener/src/System/Net/Windows/RequestContextBase.cs]: I've checked it out and am confident the dispose pattern is correctly implemented here.

**Pull request merged**